src/replay/service.zig

```diff
@@ -91,19 +96,1269 @@ fn advanceReplay(state: *ReplayState) !void {
+                ancestor_hashes_replay_update_sender,
+                .{ .kind = .dead, .slot = slot },
+            );
+        }
+    }
+
+    fn duplicate(
+        allocator: std.mem.Allocator,
+        logger: sig.trace.Logger,
+        slot: sig.core.Slot,
+        root: sig.core.Slot,
+        duplicate_slots_tracker: *DuplicateSlotsTracker,
+        fork_choice: *sig.consensus.HeaviestSubtreeForkChoice,
+        duplicate_state: DuplicateState,
+    ) !void {
+        logger.info().logf(
+            "check_slot_agrees_with_cluster() slot: {}, root: {}, slot_state_update: {}",
+            .{ slot, root, duplicate_state },
+        );
+
+        if (slot <= root) {
+            return;
+        }
+
+        // Needs to happen before the bank_frozen_hash.is_none() check below to account for duplicate
+        // signals arriving before the bank is constructed in replay.
+        if (try duplicate_slots_tracker.sorted_map.fetchPut(allocator, slot, {})) |_| {
+            // If this slot has already been processed before, return
+            return;
+        }
+
+        // datapoint_info!(
+        //     "duplicate_slot",
+        //     ("slot", slot, i64),
+        //     (
+        //         "duplicate_confirmed_hash",
+        //         duplicate_state
+        //             .duplicate_confirmed_hash
+        //             .unwrap_or_default()
+        //             .to_string(),
+        //         String
+        //     ),
+        //     (
+        //         "my_hash",
+        //         duplicate_state
+        //             .bank_status
+        //             .bank_hash()
+        //             .unwrap_or_default()
+        //             .to_string(),
+        //         String
+        //     ),
+        // );
+
+        const bank_status = duplicate_state.bank_status;
+        const duplicate_confirmed_hash = duplicate_state.duplicate_confirmed_hash;
+
+        switch (bank_status) {
+            .dead, .frozen => {},
+            // No action to be taken yet
+            .unprocessed => return,
+        }
+
+        // If the cluster duplicate_confirmed some version of this slot
+        // then either the `SlotStateUpdate::DuplicateConfirmed`, `SlotStateUpdate::BankFrozen`,
+        // or `SlotStateUpdate::Dead` state transitions will take care of marking the fork as
+        // duplicate if there's a mismatch with our local version.
+        if (duplicate_confirmed_hash == null) {
+            // If we have not yet seen any version of the slot duplicate confirmed, then mark
+            // the slot as duplicate
+            if (bank_status.bankHash()) |bank_hash| {
+                try resulting_change.markSlotDuplicate(slot, fork_choice, bank_hash);
+            }
+        }
+    }
+
+    fn epochSlotsFrozen(
+        allocator: std.mem.Allocator,
+        logger: sig.trace.Logger,
+        slot: sig.core.Slot,
+        root: sig.core.Slot,
+        epoch_slots_frozen_slots: *EpochSlotsFrozenSlots,
```

Comment:
Nit: Maybe have `epoch_slots_frozen_slots` and  `epoch_slots_frozen_state` next to each other.

Reply:
73c5ca430f4afbced10f1b83e56427b71e8944b5